src/dag_algo/mod.rs

```diff
@@ -428,6 +428,60 @@ pub fn lexicographical_topological_sort(
     Ok(PyList::new(py, out_list).into())
 }
 
+/// Return the topological generations of a DAG
+///
+/// A topological generation is node collection in which ancestors of a node in each
+/// generation are guaranteed to be in a previous generation, and any descendants of
+/// a node are guaranteed to be in a following generation. Nodes are guaranteed to
+/// be in the earliest possible generation that they can belong to.
+///
+/// :param PyDiGraph graph: The DAG to get the topological generations from
+///
+/// :returns: A list of topological generations.
+/// :rtype: list
+///
+/// :raises DAGHasCycle: if a cycle is encountered while sorting the graph
+#[pyfunction]
+#[pyo3(text_signature = "(dag, /)")]
+pub fn topological_generations(dag: &digraph::PyDiGraph) -> PyResult<Vec<NodeIndices>> {
+    if !is_directed_acyclic_graph(dag) {
```

Comment:
This check basically performs another topological sort. You can perform yours with the current code and if at the end it does not include all the nodes in the graph return that it was not a DAG

Reply:
Yeah I did not realize this, I made some change to the code based on your suggestion